File: backend/app/core/nango.py
"""
Nango OAuth Integration Client
"""
import httpx
from typing import Optional, Dict, Any
from app.config import settings

class NangoClient:
    """Client for interacting with Nango OAuth proxy"""
    
    def __init__(self):
        self.host = settings.NANGO_HOST.rstrip('/')
        self.secret_key = settings.NANGO_SECRET_KEY
    
    async def get_connection(self, connection_id: str, provider: str) -> Optional[Dict[str, Any]]:
        """Get connection details from Nango"""
        url = f"{self.host}/connection/{connection_id}"
        params = {'provider_config_key': provider}
        
        async with httpx.AsyncClient() as client:
            response = await client.get(
                url,
                headers={
                    'Authorization': f'Bearer {self.secret_key}',
                    'Content-Type': 'application/json'
                },
                params=params
            )
            
            if response.is_success:
                return response.json()
        return None
    
    async def get_access_token(self, connection_id: str, provider: str) -> Optional[str]:
        """Get access token for a connection"""
        connection = await self.get_connection(connection_id, provider)
        if connection:
            # Extract token from connection data
            credentials = connection.get('credentials', {})
            return credentials.get('access_token') or connection.get('access_token')
        return None
    
    async def proxy_request(
        self,
        connection_id: str,
        provider: str,
        method: str,
        path: str,
        json: Optional[Dict[str, Any]] = None,
        params: Optional[Dict[str, Any]] = None,
        base_url_override: Optional[str] = None
    ) -> Optional[Dict[str, Any]]:
        """Make a proxied request through Nango"""
        base_url = base_url_override or self.host
        url = f"{base_url}/proxy{path}" if not path.startswith('/proxy') else f"{base_url}{path}"
        
        headers = {
            'Authorization': f'Bearer {self.secret_key}',
            'Content-Type': 'application/json',
            'Provider-Config-Key': provider,
            'Connection-Id': connection_id
        }
        
        if base_url_override:
            headers['Base-Url-Override'] = base_url_override
        
        async with httpx.AsyncClient() as client:
            if method.upper() == 'GET':
                response = await client.get(url, headers=headers, params=params)
            elif method.upper() == 'POST':
                response = await client.post(url, headers=headers, json=json, params=params)
            elif method.upper() == 'PATCH':
                response = await client.patch(url, headers=headers, json=json, params=params)
            elif method.upper() == 'PUT':
                response = await client.put(url, headers=headers, json=json, params=params)
            elif method.upper() == 'DELETE':
                response = await client.delete(url, headers=headers)
            else:
                raise ValueError(f"Unsupported HTTP method: {method}")
            
            if response.is_success:
                content_type = response.headers.get('content-type', '')
                if 'application/json' in content_type:
                    return response.json()
                return {'status': 'success'}
            else:
                # Return error details for better debugging
                error_detail = f"HTTP {response.status_code}"
                try:
                    error_data = response.json()
                    error_detail = error_data.get('message') or error_data.get('error') or str(error_data)
                except:
                    error_detail = response.text or error_detail
                raise Exception(f"Nango proxy request failed: {error_detail}")
        return None


async def get_github_token(connection_id: str) -> Optional[str]:
    """Get GitHub access token for a connection"""
    client = NangoClient()
    return await client.get_access_token(connection_id, 'github')

File: frontend/src/lib/server/integrations/nango.ts
import crypto from 'crypto';

type NangoConnectSessionResponse = {
  data?: {
    token?: string;
    connect_link?: string;
    expires_at?: string;
  };
  error?: {
    code?: string;
    message?: string;
  };
  code?: string;
  message?: string;
  details?: Array<{ field?: string; issue?: string }>;
};

export type NangoConnectionSummary = {
  id: number;
  connection_id: string;
  provider: string;
  provider_config_key: string;
  created?: string;
  metadata?: Record<string, unknown> | null;
  tags?: Record<string, string | undefined>;
  errors?: Array<{ type?: string; log_id?: string }>;
};

type NangoConnectionsListResponse = {
  connections?: NangoConnectionSummary[];
  error?: {
    code?: string;
    message?: string;
  };
};

const NANGO_API_BASE_URL = process.env.NANGO_API_BASE_URL || 'https://api.nango.dev';

const NANGO_PROVIDER_CONFIG: Record<string, { integrationId: string; label: string }> = {
  granola: {
    integrationId:
      process.env.NANGO_GRANOLA_INTEGRATION_ID ||
      process.env.NANGO_GRANOLA_PROVIDER_CONFIG_KEY ||
      'granola',
    label: 'Granola',
  },
};

function getNangoApiKey() {
  const apiKey = process.env.NANGO_API_KEY;
  if (!apiKey) {
    throw new Error('Missing NANGO_API_KEY');
  }
  return apiKey;
}

function nangoUrl(path: string) {
  return `${NANGO_API_BASE_URL.replace(/\/$/, '')}${path}`;
}

async function parseNangoError(response: Response) {
  const data = (await response.json().catch(() => ({}))) as NangoConnectSessionResponse;
  const detail = Array.isArray(data.details)
    ? data.details
        .map((entry) => [entry.field, entry.issue].filter(Boolean).join(': '))
        .filter(Boolean)
        .join('; ')
    : '';
  const message = data.error?.message || data.message || data.error?.code || data.code;
  return [message, detail].filter(Boolean).join(' - ') || `Nango request failed with ${response.status}`;
}

export function supportedNangoProviders() {
  return Object.keys(NANGO_PROVIDER_CONFIG);
}

export function hasNangoApiKey() {
  return Boolean(process.env.NANGO_API_KEY);
}

export function nangoIntegrationForProvider(provider: string) {
  return NANGO_PROVIDER_CONFIG[provider];
}

export function providerForNangoIntegration(integrationId: string) {
  const normalizedIntegrationId = integrationId.trim().toLowerCase();
  return Object.entries(NANGO_PROVIDER_CONFIG).find(([, config]) => (
    config.integrationId.trim().toLowerCase() === normalizedIntegrationId
  ))?.[0] ?? normalizedIntegrationId;
}

export function resolveNangoWebhookUrl(origin: string) {
  if (process.env.NANGO_WEBHOOK_URL) return process.env.NANGO_WEBHOOK_URL;
  const url = new URL('/api/webhooks/nango', origin);
  if (url.hostname === 'localhost' || url.hostname === '127.0.0.1') return null;
  return url.toString();
}

export async function createNangoConnectSession(params: {
  provider: string;
  userId: string;
  userEmail?: string | null;
  userDisplayName?: string | null;
  organizationId: string;
  organizationName?: string | null;
  webhookUrl?: string | null;
}) {
  const integration = nangoIntegrationForProvider(params.provider);
  if (!integration) {
    throw new Error(`Unsupported Nango provider: ${params.provider}`);
  }

  const body: Record<string, unknown> = {
    tags: {
      end_user_id: params.userId,
      end_user_email: params.userEmail || '',
      end_user_display_name: params.userDisplayName || params.userEmail || params.userId,
      organization_id: params.organizationId,
      organization_name: params.organizationName || '',
      canon_provider: params.provider,
    },
    allowed_integrations: [integration.integrationId],
  };

  if (params.webhookUrl) {
    body.integrations_config_defaults = {
      [integration.integrationId]: {
        connection_config: {
          webhook_url: params.webhookUrl,
        },
      },
    };
  }

  const response = await fetch(nangoUrl('/connect/sessions'), {
    method: 'POST',
    headers: {
      authorization: `Bearer ${getNangoApiKey()}`,
      'content-type': 'application/json',
    },
    body: JSON.stringify(body),
  });

  if (!response.ok) {
    throw new Error(await parseNangoError(response));
  }

  const data = (await response.json()) as NangoConnectSessionResponse;
  if (!data.data?.token) {
    throw new Error('Nango did not return a connect session token');
  }

  return {
    token: data.data.token,
    connectLink: data.data.connect_link ?? null,
    expiresAt: data.data.expires_at ?? null,
    integrationId: integration.integrationId,
  };
}

export async function listNangoConnectionsForUser(params: {
  userId: string;
  organizationId: string;
}) {
  const url = new URL(nangoUrl('/connections'));
  url.searchParams.set('tags[end_user_id]', params.userId);
  url.searchParams.set('tags[organization_id]', params.organizationId);
  url.searchParams.set('limit', '100');

  const response = await fetch(url, {
    headers: {
      authorization: `Bearer ${getNangoApiKey()}`,
    },
    cache: 'no-store',
  });

  if (!response.ok) {
    throw new Error(await parseNangoError(response));
  }

  const data = (await response.json()) as NangoConnectionsListResponse;
  return data.connections ?? [];
}

export async function nangoProxyGet(params: {
  provider: string;
  connectionId: string;
  endpoint: string;
  query?: Record<string, string | number | boolean | null | undefined>;
}) {
  const integration = nangoIntegrationForProvider(params.provider);
  if (!integration) {
    throw new Error(`Unsupported Nango provider: ${params.provider}`);
  }

  const url = new URL(nangoUrl(`/proxy${params.endpoint.startsWith('/') ? params.endpoint : `/${params.endpoint}`}`));
  for (const [key, value] of Object.entries(params.query ?? {})) {
    if (value === null || value === undefined) continue;
    url.searchParams.set(key, String(value));
  }

  const response = await fetch(url, {
    headers: {
      authorization: `Bearer ${getNangoApiKey()}`,
      'provider-config-key': integration.integrationId,
      'connection-id': params.connectionId,
    },
    cache: 'no-store',
  });

  if (!response.ok) {
    throw new Error(await parseNangoError(response));
  }

  return response.json() as Promise<unknown>;
}

export async function deleteNangoConnection(params: { provider: string; connectionId: string }) {
  const integration = nangoIntegrationForProvider(params.provider);
  if (!integration) return { skipped: true };

  const url = new URL(nangoUrl(`/connections/${encodeURIComponent(params.connectionId)}`));
  url.searchParams.set('provider_config_key', integration.integrationId);

  const response = await fetch(url, {
    method: 'DELETE',
    headers: {
      authorization: `Bearer ${getNangoApiKey()}`,
    },
  });

  if (response.status === 404) return { deleted: false, missing: true };
  if (!response.ok) {
    throw new Error(await parseNangoError(response));
  }

  return { deleted: true };
}

export function verifyNangoWebhookSignature(rawBody: string, signature: string | null) {
  const signingKey = process.env.NANGO_WEBHOOK_SIGNING_KEY;
  if (!signingKey) {
    return process.env.NODE_ENV !== 'production';
  }
  if (!signature) return false;

  const expected = crypto.createHmac('sha256', signingKey).update(rawBody).digest('hex');
  const signatureBuffer = Buffer.from(signature, 'hex');
  const expectedBuffer = Buffer.from(expected, 'hex');
  if (signatureBuffer.length !== expectedBuffer.length) return false;
  return crypto.timingSafeEqual(signatureBuffer, expectedBuffer);
}
